### PR TITLE
Verb constraints

### PR DESCRIPTION
**Pull request by shans**
1) Allow particles specified by verbs to hold slot constraints. For example:

```recipe
  particle can doTheThing
    consume bar
      provide foo
```

This means that a particle or recipe can only substitute into doTheThing if it consumes a bar slot and provides a foo slot.

2) Make match-recipe-by-verb obey these constraints. Note that match-particle-by-verb doesn't, yet.

3) Ensure that in at least some cases slot connections made with these constraints persist after recipe substitution. For example:

```recipe
  particle can doTheThing
    consume foo as s1
      provide bar as s2
  P
    consume master
      provide foo as s1
  Q
    consume bar as s2
```

Will result in the matched `foo` in the recipe selected to replace doTheThing consuming `P`'s `foo`, and the matched `bar` providing a slot for `Q`'s `bar`.

This almost certainly doesn't yet work in all circumstances, for example I haven't tested what happens when the slot is shared across multiple consumers.

At some point we're going to want to do the same thing as this for handles, too.

### CODE REVIEW

File: runtime/manifest.js [hunk 1]
```diff
@@ -764,19 +765,25 @@ ${e.message}
...
+          // Validate consumed and provided slots names are according to spec.
+          if (!particle.spec.slots.has(slotConnectionItem.param)) {
             throw new ManifestError(
-                ps.location,
-                `Provided slot '${ps.param}' is not defined by '${particle.name}'`);
+                slotConnectionItem.location,
+                `Consumed slot '${slotConnectionItem.param}' is not defined by '${particle.name}'`);
           }
-        });
+          slotConnectionItem.providedSlots.forEach(ps => {
```
Comment (dstockwell) [inline runtime/manifest.js]: prefer `for (... of` over `.forEach`

File: runtime/strategies/match-recipe-by-verb.js [hunk 2]
```diff
@@ -27,20 +38,80 @@ export default class MatchRecipeByVerb extends Strategy {
...
+
+              for (let provideSlot in slotConstraints[consumeSlot].providedSlots) {
+                let slot = slotConstraints[consumeSlot].providedSlots[provideSlot];
+                let {mappedSlot} = outputRecipe.updateToClone({mappedSlot: slot});
+                for (let particle of particles) {
+                  if (particle.consumedSlotConnections[consumeSlot]) {
+                    if (particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot]) {
+                      let oldSlot = particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot];
+                      oldSlot.remove();
+                      particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot] = mappedSlot;
```
Comment (dstockwell) [inline runtime/strategies/match-recipe-by-verb.js]: This looks safe, maybe 🤷‍♂️ , but I think the API intended that we treat the `providedSlots` object as immutable.

File: runtime/manifest.js (hunk 1 above)
Comment (shans) [reply runtime/manifest.js]: I didn't actually change this...

Comment (dstockwell) [reply runtime/manifest.js]: Oh, sorry, the diff was pretty gross.

**Pull request merged**